### PR TITLE
Feature/jwt

### PR DESCRIPTION
**Pull request by bodya4243**
1) Add JWT

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -76,6 +76,27 @@
...
 			<version>2.1.0</version>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>io.jsonwebtoken</groupId>
+			<artifactId>jjwt-api</artifactId>
+			<version>0.11.2</version>
```
Comment (vovasalyha) [inline pom.xml]: It's a good practice to extract versions to properties.

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**","/swagger-ui/**")
```
Comment (vovasalyha) [inline src/main/java/com/example/onlinebookstore/config/SecurityConfig.java]: Also, add `/v3/api-docs/**`

File: src/main/java/com/example/onlinebookstore/dto/UserLoginRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.example.onlinebookstore.dto;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.Size;
+
+public record UserLoginRequestDto(
+        @NotEmpty
+        @Size(min = 8, max = 20)
```
Comment (vovasalyha) [inline src/main/java/com/example/onlinebookstore/dto/UserLoginRequestDto.java]: 20 chars for email seems too restrictive. I'm sure there are a lot of longer emails.

File: src/main/java/com/example/onlinebookstore/security/JwtUtil.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    public boolean isValidToken(String token) {
+        try {
+            Jws<Claims> claimsJws = Jwts.parserBuilder()
+                    .setSigningKey(secret)
+                    .build()
+                    .parseClaimsJws(token);
+
+            return !claimsJws.getBody().getExpiration().before(new Date());
+        } catch (JwtException | IllegalArgumentException e) {
+            throw new JwtException("Expired or invalid JWT token");
```
Comment (vovasalyha) [inline src/main/java/com/example/onlinebookstore/security/JwtUtil.java]:
```suggestion
            return false;
```

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
```
Comment (vovasalyha) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]: This number can be a bit magic if you are not in the context and looking at it the first time.

File: src/main/resources/application.properties [hunk 6]
```diff
@@ -5,6 +5,9 @@ spring.datasource.username=${DB_USER}
 spring.datasource.password=${DB_PASSWORD}
 spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
 server.port=8080
-server.servlet.context-path=/api
 spring.jpa.hibernate.ddl-auto=validate
+server.servlet.context-path=/api
 spring.jpa.show-sql=true
+
+jwt.expiration=300000
+jwt.secret=wefwqfdfwqfq4f3fq3af43453452ef34fewfewfe
```
Comment (vovasalyha) [inline src/main/resources/application.properties]: It's better to get sensitive properties from system properties and not pushing it to git.

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Looks good, but see a couple of comments.

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]: Please make string literal constants

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            return bearerToken.substring(BEARER_PREFIX_LENGTH);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring("Bearer ".length);
```

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java (hunk 2 above)
Comment (andrii-hoienko) [reply src/main/java/com/example/onlinebookstore/config/SecurityConfig.java]: I don`t see this changes

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left some minor comments

**Comment by bodya4243** [comment]
/v3/api-docs/** is already in project.
image

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader(AUTHORIZATION_HEADER);
+
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(BEARER_PREFIX)) {
+            return bearerToken.substring("Bearer ".length());
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring(BEARER_PREFIX.length());
```

**Review by karma-o — APPROVED** [review]
GJ!

**Pull request merged**